contracts/staking/contracts/src/sys/MixinFinalizer.sol

```diff
@@ -38,56 +38,54 @@ contract MixinFinalizer is
     ///      If there were no active pools in the closing epoch, the epoch
     ///      will be instantly finalized here. Otherwise, `finalizePool()`
     ///      should be called on each active pool afterwards.
-    /// @return poolsRemaining The number of unfinalized pools.
+    /// @return poolsToFinalize The number of unfinalized pools.
     function endEpoch()
         external
-        returns (uint256 poolsRemaining)
+        returns (uint256)
     {
         uint256 closingEpoch = currentEpoch;
-        IStructs.UnfinalizedState memory state = unfinalizedState;
+        uint256 prevEpoch = closingEpoch.safeSub(1);
 
         // Make sure the previous epoch has been fully finalized.
-        if (state.poolsRemaining != 0) {
+        uint256 poolsToFinalizeFromPrevEpoch = aggregatedStatsByEpoch[prevEpoch].poolsToFinalize;
+        if (poolsToFinalizeFromPrevEpoch != 0) {
             LibRichErrors.rrevert(
                 LibStakingRichErrors.PreviousEpochNotFinalizedError(
-                    closingEpoch.safeSub(1),
-                    state.poolsRemaining
+                    prevEpoch,
+                    poolsToFinalizeFromPrevEpoch
                 )
             );
         }
 
-        // Convert all ETH to WETH
+        // Since it is finalized, we no longer need stats for the previous epoch.
+        delete aggregatedStatsByEpoch[prevEpoch];
```

Comment:
Yeah, I was thinking about that too. If we do remove this then how would you feel about also retaining stats for individual pools, which could be more relevant to dapps. 

It was previously beneficial to clear a pool's stats to save gas when implicitly finalizing during un/delegating and withdrawals. Since we're no longer longer doing implicit finalizations this may be a non-issue.

Reply:
Note: if we retain pool stats then we'd have to add a new field to the pool struct to indicate whether it has been finalized.

Reply:
I'd prefer to delete those since the number of pools to finalize can theoretically be up to 10mm (with the current min stake).